src/component.ts

```diff
@@ -1,4 +1,36 @@
+import { BitSet } from 'bitset';
+import { Constructor } from './world';
+
+function* bitmaskGenerator(): IterableIterator<BitSet> {
+  let n = 0;
+
+  while (true) {
+    const mask = new BitSet(0);
+    mask.set(n);
+    ++n;
+
+    yield mask;
+  }
+}
+
+export interface IComponent {
+  readonly bitmask: BitSet;
+}
+
+export type ComponentConstructor = Constructor<Component> & IComponent;
+
 export abstract class Component {
+  protected static readonly _bitmaskGenerator = bitmaskGenerator();
+  protected static _bitmask: BitSet | null = null;
+
   // tslint:disable-next-line: variable-name
   protected readonly __component = true;
+
+  public static get bitmask() {
+    if (this._bitmask == null) {
+      this._bitmask = this._bitmaskGenerator.next().value;
+    }
```

Comment:
Should this be in the constructor instead of lazily evaluated?

Reply:
We don't set this in the constructor because we want `bitmask` to be a static property on Component constructors. We can't set static fields in the `Component` constructor.